rbc/omniscidb.py

```diff
@@ -212,6 +212,95 @@ def get_client_config(**config):
     return config
 
 
+def is_udtf(sig):
+    """Check if signature is a table function signature.
+    """
+    if sig[0].annotation().get('kind') == 'UDTF':
+        return True
+    for a in sig[1]:
+        if isinstance(a, (OmnisciOutputColumnType, OmnisciColumnType,
+                          OmnisciColumnListType, OmnisciTableFunctionManagerType)):
+            return True
+    return False
+
+
+output_buffer_sizer_map = dict(
+    ConstantParameter='kUserSpecifiedConstantParameter',
+    RowMultiplier='kUserSpecifiedRowMultiplier',
+    Constant='kConstant',
+    SpecifiedParameter='kTableFunctionSpecifiedParameter',
+    PreFlight='kPreFlightParameter')
+output_buffer_sizer_map[None] = output_buffer_sizer_map['RowMultiplier']
+
+user_specified_output_buffer_sizers = {
+    'kUserSpecifiedConstantParameter', 'kUserSpecifiedRowMultiplier',
+}
+
+
+def type_to_type_name(typ: typesystem.Type):
+    """Return typesystem.Type as DatumType name.
+    """
+    styp = typ.tostring(use_annotation=False, use_name=False)
+    type_name = dict(
+        int8='TINYINT',
+        int16='SMALLINT',
+        int32='INT',
+        int64='BIGINT',
+        float32='FLOAT',
+        float64='DOUBLE',
+    ).get(styp)
+    if type_name is not None:
+        return type_name
+    raise NotImplementedError(f'converting `{styp}` to DatumType not supported')
+
+
+def type_name_to_dtype(type_name):
+    """Return DatumType name as the corresponding numpy dtype.
+    """
+    dtype = dict(
+        SMALLINT=numpy.int16,
+        INT=numpy.int32,
+        BIGINT=numpy.int64,
+        FLOAT=numpy.float32,
+        # DECIMAL=,
+        DOUBLE=numpy.float32,
+        STR=numpy.str0,
+        # TIME=,
+        # TIMESTAMP=,
+        # DATE=,
+        BOOL=numpy.bool8,
+        # INTERVAL_DAY_TIME=,
+        # INTERVAL_YEAR_MONTH=,
+        # POINT=,
+        # LINESTRING=,
+        # POLYGON=,
+        # MULTIPOLYGON=,
+        TINYINT=numpy.int8,
+        # GEOMETRY=,
+        # GEOGRAPHY=
+    ).get(type_name)
+    if dtype is not None:
+        return dtype
+    raise NotImplementedError(
+        f'convert DatumType `{type_name}` to numpy dtype')
+
+
+class Query(RemoteCallCapsule):
```

Comment:
It is a good suggestion. Since the object is exposed to end-users, it's str/repr representations ought to be informative:
Suggestion:
```python
>>> str(incr(5)))
'SELECT incr(CAST(5 AS BIGINT))'
>>> repr(incr(5))
OmniciQueryCapsule('SELECT incr(CAST(5 AS BIGINT))')
```
@guilhermeleobas @tupui what do you think?

Reply:
I like the proposal.

Another question I am having is: do we want to cast the input? This might confuse some user no? If we have int to float it's fine but in some case we will have float to int and without any warning the result can introduce bug in users' code.

Reply:
I agree that some casts can be eliminated, but not all. For instance, `incr(np.int64(5))` should result in calling int64 version of `incr`, that is, `str(np.int64(5))` should lead to `'SELECT incr(CAST(5 AS BIGINT))'`.
On the other hand, `str(np.int32(5))` could be `SELECT incr(5)`.

Using CAST depends on how omniscidb interprets literal values and to drop some casts, RBC ought to record the casting rules of omniscidb (which are still in flux). Since we don't have this recorded, explicit CAST is better than no CAST.

Reply:
Query is now OmnisciQueryCapsule

Reply:
> in some case we will have float to int and without any warning the result can introduce bug in users code ..

Not possible: the casts are generated based on user inputs and on available implementations. If there is a mismatch in input types and expected argument types, a TypeError will be raised. Try, for instance, `arange(5.5)` that will fail because `arange` expects integer input.

Reply:
Oh great 👍